course5/homework6/string.cc

```diff
@@ -0,0 +1,115 @@
+#include <memory>
+#include <cstring>
+
+#include "string.h"
+
+namespace jds {
+
+String::String()
+{
+	mLength = 0;
+}
+
+String::String(String const & that)
+{
+	mLength = that.mLength;
+	mBuffer = std::make_unique<char[]>(mLength + 1);
+	std::memcpy(this->mBuffer.get(), that.mBuffer.get(), that.mLength);
+	mBuffer.get()[mLength] = '\0';
+}
+
+String::String(String && that)
+{
+	mBuffer = std::unique_ptr<char []>(that.mBuffer.release());
+	mLength = that.mLength;
+	that.mLength = 0;
+}
+
+String::String(char const * source)
+{
+	mLength = strlen(source);
+	mBuffer = std::make_unique<char[]>(mLength + 1);
+	std::memcpy(mBuffer.get(), source, mLength);
+	mBuffer.get()[mLength] = '\0';
+}
+
+void String::operator=(String const & that)
+{
+	mLength = that.mLength;
+	mBuffer = std::make_unique<char[]>(mLength + 1);
+	std::memcpy(mBuffer.get(), that.mBuffer.get(), mLength);
+	mBuffer.get()[mLength] = '\0';
+}
+
+void String::operator=(String && that)
+{
+	mBuffer = std::unique_ptr<char []>(that.mBuffer.release());
+	mLength = that.mLength;
+	that.mLength = 0;
+}
+
+void String::operator=(char const * source)
+{
+	mLength = strlen(source);
+	mBuffer = std::make_unique<char[]>(mLength + 1);
+	std::memcpy(mBuffer.get(), source, mLength);
+	mBuffer.get()[mLength] = '\0';
+}
+
+unsigned String::length() const
+{
+	return mLength;
+}
+
+char& String::operator[](unsigned index)
+{
+	if (index >= mLength)
```

Comment:
A string with length x has only valid indices 0 through x-1, hence if index == length it's already out of range. If the string is default constructed it's empty and has no accessible characters.

If you don't like >= and <= the logic could be inverted to operate normally when index < mLength and error out in the other branch.